### DISCUSSION TITLE
Arduino DUE SAM3X8E Ramps1.6 spindle PWM output

### DISCUSSION THREAD

**Discussion by MrSpindle** [terjeio/grblHAL#238] [Q&A]
Hi everybody,
total newby here who wants to switch from grbl on an Arduino Mega with ramps 1.4 to Arduino Due grblHAL and ramps1.6.
I was able to compile the grblHAL in Arduino IDE with compatibility mode to work with UGS, that worked well- I can also switch on the spindle wit M3 Sxxx commands and switch it off with M5. But I cannot find the PWM signal, on the Arduino Due. Spindle Enable Pin is on Digital Pin 4 (works if I enter M3 S200 ). The Ramps is not powered up so the Due is only connected to USB. 
Pin 11, 12 13 do not work
Any ideas, any help would be musch appreciated - thanks!

the Readme from grblHAL is referring to this build:
Latest build date is 20201103

$0 | 10.0 | Step pulse time
-- | -- | --
$1 | 255 | Step idle delay
$2 | 0 | Step pulse invert
$3 | 3 | Step direction invert
$4 | 0 | Invert step enable pin
$5 | 0 | Invert limit pins
$6 | 0 | Invert probe pin
$10 | 1 | Status report options
$11 | 0.010 | Junction deviation
$12 | 0.002 | Arc tolerance
$13 | 0 | Report in inches
$14 | 15 |  
$15 | 0 |  
$16 | 0 |  
$17 | 0 |  
$18 | 0 |  
$19 | 0 |  
$20 | 0 | Soft limits enable
$21 | 0 | Hard limits enable
$22 | 0 | Homing cycle enable
$23 | 0 | Homing direction invert
$24 | 25.000 | Homing locate feed rate
$25 | 500.000 | Homing search seek rate
$26 | 250 | Homing switch debounce delay
$27 | 1.000 | Homing switch pull-off distance
$28 | 0.100 |  
$29 | 0.0 |  
$30 | 256. | Maximum spindle speed
$31 | 0. | Minimum spindle speed
$32 | 0 | Laser-mode enable
$33 | 5000.000 |  
$34 | 0.000 |  
$35 | 0.000 |  
$36 | 100.000 |  
$37 | 0 |  
$39 | 0 |  
$40 | 0 |  
$41 | 0 |  
$42 | 2 |  
$43 | 1 |  
$44 | 4 |  
$45 | 3 |  
$46 | 0 |  
$56 | 5.000 |  
$57 | 100.000 |  
$58 | -5.000 |  
$59 | 500.000 |  
$60 | 0 |  
$61 | 0 |  
$62 | 0 |  
$63 | 2 |  
$64 | 0 |  
$65 | 0 |  
$100 | 23.600 | X-axis travel resolution
$101 | 13.400 | Y-axis travel resolution
$102 | 8.889 | Z-axis travel resolution
$110 | 7000.000 | X-axis maximum rate
$111 | 5000.000 | Y-axis maximum rate
$112 | 5000.000 | Z-axis maximum rate
$120 | 150.000 | X-axis acceleration
$121 | 150.000 | Y-axis acceleration
$122 | 50.000 | Z-axis acceleration
$130 | 400.000 | X-axis maximum travel
$131 | 2300.000 | Y-axis maximum travel
$132 | 360.000 | Z-axis maximum travel

**Comment by MrSpindle** [comment]
![M3 S120 not working Spindle Speed](https://user-images.githubusercontent.com/78793836/107358575-d9398000-6ad3-11eb-9aa5-03f8c4b36b7a.PNG)

I can enter the speed but do not get any status on the speed- it is working with the old Mega 5x setup though

![M3 S120  working Spindle Speed on MEGA 2560](https://user-images.githubusercontent.com/78793836/107358806-27e71a00-6ad4-11eb-8451-9a007ae66cb6.PNG)

**Comment by MrSpindle** [comment]
Ok I changed the switched the digital output pins for the spindle direction and pwm in the ramps_1.6_map.h and now the pwm output works on pin 5. UGS still does not show me an spindle speed. 
ramps_1.6_map.h Lines 113 to 130 

#ifndef MODBUS_ENABLE

// Define spindle enable and spindle direction output pins.
#define SPINDLE_ENABLE_PORT     PIOC
#define SPINDLE_ENABLE_PIN      26  // Due Digital Pin 4
#define SPINDLE_ENABLE_BIT      (1<<SPINDLE_ENABLE_PIN)
#define SPINDLE_DIRECTION_PORT  PIOC
#define SPINDLE_DIRECTION_PIN   22  // was25 Due Digital Pin 8 was 5
#define SPINDLE_DIRECTION_BIT   (1<<SPINDLE_DIRECTION_PIN)

// Start of PWM & Stepper Enabled Spindle
#define SPINDLE_PWM_TIMER   (TC2->TC_CHANNEL[0])
#define SPINDLE_PWM_CCREG   2
#define SPINDLE_PWM_PORT    PIOC
#define SPINDLE_PWM_PIN     25  // was22 Due Digital Pin 5 was  8 // PWML5
#define SPINDLE_PWM_BIT     (1<<SPINDLE_PWM_PIN)

#endif

**Reply by terjeio** [reply]
> Ok I changed the switched the digital output pins for the spindle direction and pwm in the ramps_1.6_map.h and now the pwm output works on pin 5.

I'll have to go back and check this - the code does not agree with the [board layout file](https://github.com/terjeio/grblHAL/blob/master/drivers/SAM3X8E/Arduino%20Due%20shield%20pin%20mappings/ramps_1.6_map.md) either.

> UGS still does not show me an spindle speed.

What is the output from a status request, `?`, when the spindle is running?

 It could be a bug in the driver, I'll check that later when I have finished some other work.
BTW, did you use the test branch for your build?

**Reply by terjeio** [reply]
> UGS still does not show me an spindle speed.

The $10 setting has many new flags added to it that is not visible/possible to change when compatibility mode is > 1:
```
$10: Status report options as bitfield:
    0 - Position in machine coordinate (1)
    1 - Buffer state (2)
    2 - Line numbers (4)
    3 - Feed & speed (8)
    4 - Pin state (16)
    5 - Work coordinate offset (32)
    6 - Overrides (64)
    7 - Probe coordinates (128)
    8 - Buffer sync on WCO change (256)
    9 - Parser state (512)
    10 - Alarm substatus (1024)
    11 - Run substatus (2048)
```
If bit 3 is not set spindle speed is not added to the status report. Some options to try for fixing this:

1. Compile/upload with compatibilty level set to 0 and set `$10=511`, then revert back to the previous compatibility level.
2. Reset all settings to default with `$RST=*`, this should set hidden bits back to default. This assumes you have not removed the comment from `DEFAULT_NO_REPORT_CURRENT_FEED_SPEED` in _grbl/config.h_.
3. Update to the [test branch](https://github.com/terjeio/grblHAL/tree/test).

**Reply by MrSpindle** [reply]
ok thank you very much for your answer, the spindle speed status is not super important as long as I can control it through g-code.
But there is another issue-should I open another thread for that? 
The Z axis direction is only going into one direction- the pin 48 on the Arduino is not going high with 3volts it stays on 0 volt. 
I tried other drivers and axis (x and y) and this only appears on the z axis. I also had to disconnect the enable pin. 
I am using external stepper drivers. I have checked the ramps 1.6 traces from the arduino-pins to the stepper driver sockets and they are connected. 
So I am assuming there might be a mix up with the pins in the mapping file? 

testing the test branch now

**Reply by MrSpindle** [reply]
ok we solved the Z axis issue- there is a typo in ramps_1.6_map.h in line 58 
#define Z_DIRECTION_BIT     (1<<Z_STEP_PIN) 
should be 
#define Z_DIRECTION_BIT     (1<<Z_DIRECTION_PIN)

sorry I don't know where to report, we are not so familiar with git here...

**Reply by terjeio** [reply]
I have found it too, no need to report. There are copy/paste errors in other map files too.

**Reply by MrSpindle** [reply]
ok perfect, we are using the test branch right now. apparently the $G soft reset ist not working as well. I have to reset the Due to reset the grbl. 
Using UGS with compatibility mode still.

**Reply by terjeio** [reply]
> apparently the $G soft reset ist not working

`$G` is for outputting a parser state report, use `<CTRL>+<X>` or `<CTRL>+<C>` for soft reset.

**Reply by MrSpindle** [reply]
ok, I may have mixed stuff up..
 I had to set $10=0 in order to zero all axis- I suspect that grbl was set to zero but UGS didn't show that. 
Thanks again for your support and work on grbl!!!
